src/tqec/compile/observables/builder.py

```diff
@@ -0,0 +1,319 @@
+    cube is only responsible for the stabilizer measurements within its
+    bounding box. Collecting the measurements from the two cubes connected by
+    the pipe will give the full stabilizer measurements, of which the product
+    determines the parity of the logical operators.
+
+    This function calculates the coordinates of the measurement qubits
+    in the local coordinate system of the cube.
+    """
+    stabilizers: list[tuple[float, float]] = []
+    # We calculate the qubits for the connect_to=SignedDirection3D(Direction3D.X, True) case
+    # and rotate to get the correct orientation.
+    for i in range(cube_shape.x // 2):
+        x = cube_shape.x - i - 0.5
+        for j in range(cube_shape.y // 2):
+            y = (1 - i % 2) + 2 * j + 0.5
+            stabilizers.append((x, y))
+    # rotate all coordinates around the block center:
+    # rx = cx + a * (x - cx) - b * (y - cy)
+    # ry = cy + b * (x - cx) + a * (y - cy)
+    # in which (cx, cy) is the center of the block, a = cos(theta), b = sin(theta)
+    # and theta is the angle of rotation.
+    a, b = 1, 0
+    match connect_to:
+        case SignedDirection3D(Direction3D.X, False):
+            a, b = -1, 0
+        case SignedDirection3D(Direction3D.Y, True):
+            a, b = 0, 1
+        case SignedDirection3D(Direction3D.Y, False):
+            a, b = 0, -1
+        case _:
+            pass
+    cx, cy = cube_shape.x // 2, cube_shape.y // 2
+    return [
+        (
+            cx + a * (x - cx) - b * (y - cy),
+            cy + b * (x - cx) + a * (y - cy),
+        )
+        for x, y in stabilizers
+    ]
+
+
+def _get_top_readout_spatial_junction_qubits(
+    junction_shape: Shape2D,
+    arms: JunctionArms,
+) -> list[tuple[int, int]]:
+    """The data qubits at the spatial junctions will be read out and included
+    in the logical observable.
+
+    This function calculates the coordinates of the data qubits in the
+    local coordinate system of the junction based on which arms the
+    correlation surface touches.
+    """
+    assert len(arms) == 2
+    half_x, half_y = junction_shape.x // 2, junction_shape.y // 2
+    # could use match expressions here
+    # but not sure how is `|` operator evaluated in
+    # `case JunctionArms.LEFT | JunctionArms.RIGHT:`
+    # Is that used as case separator or bitwise OR?
+    if arms == JunctionArms.LEFT | JunctionArms.RIGHT:
+        return [(x, half_y) for x in range(junction_shape.x + 1)]
+    elif arms == JunctionArms.UP | JunctionArms.DOWN:
+        return [(half_x, y) for y in range(junction_shape.y + 1)]
+    elif arms == JunctionArms.LEFT | JunctionArms.UP:
+        return [(x, half_y) for x in range(half_x)] + [
+            (half_x, y) for y in range(half_y)
+        ]
+    elif arms == JunctionArms.DOWN | JunctionArms.RIGHT:
+        return [(x, half_y) for x in range(junction_shape.x, half_x, -1)] + [
+            (half_x, y) for y in range(junction_shape.y, half_y, -1)
+        ]
```

Comment:
```suggestion
        return [(half_x + x, half_y) for x in range(half_x)] + [
            (half_x, half_y + y) for y in range(half_y)
        ]
```
should be the same because the shapes are always even numbers. The above version is easier to read for me.

Reply:
The above suggestion seems to contain off-by-1 errors, not sure it will be easier to read than the original code after fixing these.

Reply:
Both the two don't seem that complex, so I'll keep the current implementation.